src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php

```diff
@@ -48,24 +49,56 @@ public function vote(TokenInterface $token, $subject, array $attributes)
             }
 
             // as soon as at least one attribute is supported, default is to deny access
-            $vote = self::ACCESS_DENIED;
+            $vote = $this->deny();
 
-            if ($this->voteOnAttribute($attribute, $subject, $token)) {
+            $decision = $this->voteOnAttribute($attribute, $subject, $token);
+            if (\is_bool($decision)) {
+                trigger_deprecation('symfony/security-core', '5.3', 'Returning a boolean in "%s::voteOnAttribute()" is deprecated, return an instance of "%s" instead.', static::class, Vote::class);
+                $decision = $decision ? $this->grant() : $this->deny();
+            }
+
+            if ($decision->isGranted()) {
                 // grant access as soon as at least one attribute returns a positive response
-                return self::ACCESS_GRANTED;
+                return $decision;
             }
+
+            $vote->setReason($vote->getReason().trim(' '.$decision->getReason()));
         }
 
         return $vote;
     }
 
+    /**
+     * Creates a granted vote.
+     */
+    public function grant(string $reason = '', array $parameters = []): Vote
```

Comment:
Why are those public methods ? They look like helpers for the child class, and so should be protected

Reply:
I will make the protected.